compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -225,6 +226,171 @@ impl<'tcx> AbstractConst<'tcx> {
+    }
+
+    // Recursively applies substs to leaves. Also ensures that the strings corresponding
+    // to nodes in the `AbstractConst` tree are wrapped in curly braces after a substitution
+    // was applied. E.g in `{ 3 * M}` where we have substs `{ N + 2 }` for M, we want to
+    // make sure that the node corresponding to `M` is wrapped in curly braces: `{ 3 * { N +2 }}`
+    #[instrument(skip(tcx), level = "debug")]
+    fn recursive_try_print_with_replacing_substs(
+        mut self,
+        tcx: TyCtxt<'tcx>,
+    ) -> Option<PrintStyle> {
+        let mut applied_subst = false;
+
+        // try to replace substs
+        while let abstract_const::Node::Leaf(ct) = self.root(tcx) {
+            match AbstractConst::from_const(tcx, ct) {
+                Ok(Some(act)) => {
+                    applied_subst = true;
+                    self = act;
+                }
+                Ok(None) => break,
+                Err(_) => bug!("should be able to create AbstractConst here"),
+            }
+        }
+
+        debug!("after applying substs: {:#?}", self);
+        debug!("root: {:?}", self.root(tcx));
+        debug!(?applied_subst);
+
+        // How we infer precedence of operations:
+        // We use `PrintStyle::Braces(op_precedence, str)` to indicate that resulting
+        // string of recursive call may need to be wrapped in curly braces. If the
+        // precedence of the op of a node is larger than `op_precedence` of the returned
+        // subnode we need to use curly braces.
+        // If the precedence isn't larger we still send `PrintStyle::Braces` upwards
+        // since the string that includes the node which requires precedence might
+        // still require a higher-level node string to be wrapped in curly braces, e.g
+        // in `{ 3 + {M + K}}` where `M` is substituted by `N + 2` we want the result string
+        // to be `{ M + { N + 2 + K}}`.
+        match self.root(tcx) {
+            abstract_const::Node::Leaf(ct) => match ct.val {
+                ty::ConstKind::Error(_) | ty::ConstKind::Infer(_) => return None,
+                ty::ConstKind::Param(c) => {
+                    let s = format!("{}", c);
+
+                    if applied_subst {
+                        return Some(PrintStyle::Braces(None, s));
+                    } else {
+                        return Some(PrintStyle::NoBraces(s));
+                    }
+                }
+                ty::ConstKind::Value(ConstValue::Scalar(scalar)) => match scalar.to_i64() {
+                    Ok(s) => {
+                        let s = format!("{}", s);
+
+                        if applied_subst {
+                            return Some(PrintStyle::Braces(None, s));
+                        } else {
+                            return Some(PrintStyle::NoBraces(s));
+                        }
+                    }
+                    Err(_) => return None,
+                },
```

Comment:
This is the cause of the CI failures. the CI is failing on 32 bit stuff which means if we have a `usize` scalar here and call `.to_i64` we get an `InterpErrorInfo(InterpErrorInfoInner { kind: scalar size mismatch: expected 8 bytes but got 4 bytes instead, backtrace: None })` from this call but on 64bit it works fine because the sizes match.

the docs for `to_i64` say "Fails if the scalar is a pointer" but I guess that is wrong .-.
this code would also just be wrong if the scalar was a u128 or something since those values cant all fit in a i128

not sure exactly what the right thing to replace this with is but we surely have code somewhere for printing this out

Reply:
Thanks, that's a great catch. I'll try to figure out how to handle this correctly here.